File: bindings/jvm/src/main/java/io/vinarytree/libdictenstein/CollectionTraversalProfile.java
package io.vinarytree.libdictenstein;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.OptionalLong;
import io.vinarytree.interop.DictionaryEntry;
import io.vinarytree.interop.DictionaryEntryIterator;
import io.vinarytree.interop.DictionarySnapshot;
import io.vinarytree.interop.DictionaryUnitDomain;

/** Deterministic benchmark entrypoint over the public JVM collection facade. */
public final class CollectionTraversalProfile {
    private static final String SCHEMA = "libdictenstein.host-collection-traversal.v1";
    private static final int KEY_UNITS = 38;

    /** One schema-compatible machine-readable result. */
    public record Result(
            String runtime,
            String arm,
            int dictionaryEntries,
            int consumedEntriesPerPass,
            int passes,
            int warmupPasses,
            Integer batchSize,
            Integer earlyCancel,
            long elapsedNs,
            long checksum) {
        /** Encode without a JSON dependency; checksum is emitted as an unsigned JSON integer. */
        public String toJson() {
            return "{\"schema\":\"" + SCHEMA
                    + "\",\"runtime\":\"" + runtime
                    + "\",\"arm\":\"" + arm
                    + "\",\"dictionary_entries\":" + dictionaryEntries
                    + ",\"consumed_entries_per_pass\":" + consumedEntriesPerPass
                    + ",\"passes\":" + passes
                    + ",\"warmup_passes\":" + warmupPasses
                    + ",\"batch_size\":" + (batchSize == null ? "null" : batchSize)
                    + ",\"early_cancel\":" + (earlyCancel == null ? "null" : earlyCancel)
                    + ",\"elapsed_ns\":" + elapsedNs
                    + ",\"checksum\":" + Long.toUnsignedString(checksum) + "}";
        }
    }

    private record Config(
            String arm, int entries, int passes, int warmupPasses, int batchSize, int earlyCancel) {
        int consumed() { return arm.equals("stream-cancel") ? Math.min(entries, earlyCancel) : entries; }
    }
    private record CorpusEntry(byte[] key, long value) {}
    private record Drain(long checksum, int count) {}

    private CollectionTraversalProfile() {}

    /** Run one arm. Construction and warmup are excluded from {@link Result#elapsedNs()}. */
    public static Result run(String... arguments) {
        Config config = parse(arguments);
        List<CorpusEntry> corpus = corpus(config.entries);
        long expected = expectedChecksum(corpus, config.consumed());
        try (DynamicDawg dictionary = new DynamicDawg(UnitDomain.BYTE)) {
            List<Dictionary.TextEntry> mutations = corpus.stream()
                    .map(entry -> new Dictionary.TextEntry(entry.key, OptionalLong.of(entry.value)))
                    .toList();
            long inserted = dictionary.putAllBytes(mutations);
            if (inserted != corpus.size()) {
                throw new IllegalStateException("generated corpus was not inserted exactly once");
            }
            for (int pass = 0; pass < config.warmupPasses; pass++) {
                verify(drain(dictionary, config), config.consumed(), expected, "warmup");
            }
            long started = System.nanoTime();
            long checksum = 0;
            for (int pass = 0; pass < config.passes; pass++) {
                Drain result = drain(dictionary, config);
                verify(result, config.consumed(), expected, "timed drain");
                checksum += result.checksum;
            }
            long elapsed = Math.max(1L, System.nanoTime() - started);
            if (checksum != expected * config.passes) {
                throw new IllegalStateException("aggregate checksum mismatch");
            }
            return new Result(
                    "jvm-java", config.arm, config.entries, config.consumed(), config.passes,
                    config.warmupPasses, config.arm.equals("materialized") ? null : config.batchSize,
                    config.arm.equals("stream-cancel") ? config.earlyCancel : null,
                    elapsed, checksum);
        }
    }

    /** CLI entrypoint. */
    public static void main(String[] arguments) {
        try {
            System.out.println(run(arguments).toJson());
        } catch (RuntimeException failure) {
            System.err.println(failure.getMessage());
            System.exit(2);
        }
    }

    private static Config parse(String[] arguments) {
        String arm = null;
        int entries = 65_536, passes = 1, warmup = 1, batch = 256, early = 64;
        if ((arguments.length & 1) != 0) throw new IllegalArgumentException("every option requires a value");
        for (int index = 0; index < arguments.length; index += 2) {
            String option = arguments[index], value = arguments[index + 1];
            switch (option) {
                case "--arm" -> arm = value;
                case "--entries" -> entries = positive(value, option, false);
                case "--passes" -> passes = positive(value, option, false);
                case "--warmup-passes" -> warmup = positive(value, option, true);
                case "--batch-size" -> batch = positive(value, option, false);
                case "--early-cancel" -> early = positive(value, option, false);
                default -> throw new IllegalArgumentException("unknown argument: " + option);
            }
        }
        if (!List.of("materialized", "stream", "stream-cancel").contains(arm)) {
            throw new IllegalArgumentException("--arm must be materialized, stream, or stream-cancel");
        }
        Math.multiplyExact(batch, KEY_UNITS);
        return new Config(arm, entries, passes, warmup, batch, early);
    }

    private static int positive(String value, String option, boolean allowZero) {
        int result;
        try { result = Integer.parseInt(value); }
        catch (NumberFormatException failure) { throw new IllegalArgumentException(option + " must be an integer"); }
        if (allowZero ? result < 0 : result <= 0) {
            throw new IllegalArgumentException(option + " must be " + (allowZero ? "nonnegative" : "positive"));
        }
        return result;
    }

    private static List<CorpusEntry> corpus(int size) {
        List<CorpusEntry> result = new ArrayList<>(size);
        for (int index = 0; index < size; index++) {
            String key = String.format("collection/%04x/%08x/shared-suffix", index & 0x0fff, index);
            result.add(new CorpusEntry(key.getBytes(StandardCharsets.US_ASCII), index));
        }
        return result;
    }

    private static long expectedChecksum(List<CorpusEntry> corpus, int limit) {
        List<CorpusEntry> ordered = new ArrayList<>(corpus);
        ordered.sort(Comparator.comparing(CorpusEntry::key, Arrays::compareUnsigned));
        long result = 0;
        for (int index = 0; index < limit; index++) {
            CorpusEntry entry = ordered.get(index);
            result += entry.key.length ^ entry.value;
        }
        return result;
    }

    private static Drain drain(DynamicDawg dictionary, Config config) {
        if (config.arm.equals("materialized")) {
            DictionarySnapshot snapshot = dictionary.snapshot();
            long checksum = 0;
            for (DictionaryEntry entry : snapshot) checksum += checksum(entry);
            return new Drain(checksum, snapshot.size());
        }
        long checksum = 0;
        int count = 0;
        try (DictionaryEntryIterator stream = dictionary.openEntryStream(config.batchSize)) {
            while (count < config.consumed() && stream.hasNext()) {
                checksum += checksum(stream.next());
                count++;
            }
            if (config.arm.equals("stream") && (count != config.consumed() || stream.hasNext())) {
                throw new IllegalStateException("stream cardinality differs from the generated corpus");
            }
        }
        return new Drain(checksum, count);
    }

    private static long checksum(DictionaryEntry entry) {
        if (entry.key().domain() != DictionaryUnitDomain.BYTE) {
            throw new IllegalStateException("benchmark expected byte-domain entries");
        }
        return entry.key().unitCount()
                ^ entry.value().map(io.vinarytree.interop.UnsignedLong::bits).orElse(0L);
    }

    private static void verify(Drain result, int count, long checksum, String phase) {
        if (result.count != count || result.checksum != checksum) {
            throw new IllegalStateException(phase + " checksum or cardinality mismatch");
        }
    }
}

File: bindings/javascript/benchmarks/collection-traversal-profile.mjs
const SCHEMA = "libdictenstein.host-collection-traversal.v1";
const KEY_UNITS = 38;
const U64_MASK = (1n << 64n) - 1n;
const encoder = new TextEncoder();

function positive(value, name, allowZero = false) {
  if (!Number.isSafeInteger(value) || (allowZero ? value < 0 : value <= 0)) {
    throw new TypeError(`${name} must be ${allowZero ? "nonnegative" : "positive"}`);
  }
  return value;
}

export function makeCorpus(size) {
  positive(size, "entries");
  return Array.from({ length: size }, (_, index) => ({
    key: encoder.encode(`collection/${(index & 0x0fff).toString(16).padStart(4, "0")}/${index.toString(16).padStart(8, "0")}/shared-suffix`),
    value: BigInt(index),
  }));
}

function compareBytes(left, right) {
  const count = Math.min(left.length, right.length);
  for (let index = 0; index < count; index += 1) {
    if (left[index] !== right[index]) return left[index] - right[index];
  }
  return left.length - right.length;
}

function checksumEntry([key, value]) {
  if (!(key instanceof Uint8Array)) throw new TypeError("benchmark expected byte-domain entries");
  return BigInt(key.byteLength) ^ (value ?? 0n);
}

export function expectedChecksum(corpus, limit) {
  return corpus.slice().sort((left, right) => compareBytes(left.key, right.key)).slice(0, limit)
    .reduce((total, entry) => (total + (BigInt(entry.key.byteLength) ^ entry.value)) & U64_MASK, 0n);
}

function drain(dictionary, config) {
  const limit = config.arm === "stream-cancel"
    ? Math.min(config.entries, config.earlyCancel)
    : config.entries;
  if (config.arm === "materialized") {
    const snapshot = dictionary.snapshot();
    let checksum = 0n;
    for (const entry of snapshot) checksum = (checksum + checksumEntry(entry)) & U64_MASK;
    return { checksum, count: snapshot.size };
  }

  const stream = dictionary.streamEntries();
  let checksum = 0n;
  let count = 0;
  try {
    while (count < limit) {
      const batch = stream.nextBatch(Math.min(config.batchSize, limit - count));
      if (batch.length === 0) break;
      for (const entry of batch) checksum = (checksum + checksumEntry(entry)) & U64_MASK;
      count += batch.length;
    }
    if (config.arm === "stream" && (count !== limit || stream.nextBatch(config.batchSize).length !== 0)) {
      throw new Error("stream cardinality differs from the generated corpus");
    }
    return { checksum, count };
  } finally {
    stream.close();
  }
}

/** Run one browser, native Node, or Node-WASI facade arm and return its machine row. */
export function runCollectionTraversalProfile(namespace, options = {}) {
  const config = {
    arm: options.arm,
    entries: positive(options.entries ?? 65_536, "entries"),
    passes: positive(options.passes ?? 1, "passes"),
    warmupPasses: positive(options.warmupPasses ?? 1, "warmupPasses", true),
    batchSize: positive(options.batchSize ?? 256, "batchSize"),
    earlyCancel: positive(options.earlyCancel ?? 64, "earlyCancel"),
  };
  if (!["materialized", "stream", "stream-cancel"].includes(config.arm)) {
    throw new TypeError("arm must be materialized, stream, or stream-cancel");
  }
  const corpus = makeCorpus(config.entries);
  const consumed = config.arm === "stream-cancel"
    ? Math.min(config.entries, config.earlyCancel)
    : config.entries;
  const expected = expectedChecksum(corpus, consumed);
  const dictionary = namespace.dynamicDawg("byte");
  try {
    for (const entry of corpus) {
      if (!dictionary.put(entry.key, entry.value)) throw new Error("generated corpus contains a duplicate key");
    }
    for (let pass = 0; pass < config.warmupPasses; pass += 1) {
      const result = drain(dictionary, config);
      if (result.count !== consumed || result.checksum !== expected) throw new Error("warmup checksum or cardinality mismatch");
    }
    const started = performance.now();
    let checksum = 0n;
    for (let pass = 0; pass < config.passes; pass += 1) {
      const result = drain(dictionary, config);
      if (result.count !== consumed || result.checksum !== expected) throw new Error("timed checksum or cardinality mismatch");
      checksum = (checksum + result.checksum) & U64_MASK;
    }
    const elapsedNs = Math.max(1, Math.round((performance.now() - started) * 1_000_000));
    if (checksum !== (expected * BigInt(config.passes) & U64_MASK)) throw new Error("aggregate checksum mismatch");
    if (checksum > BigInt(Number.MAX_SAFE_INTEGER)) throw new RangeError("checksum exceeds the JSON safe-integer range");
    return {
      schema: SCHEMA,
      runtime: options.runtime ?? "javascript",
      arm: config.arm,
      dictionary_entries: config.entries,
      consumed_entries_per_pass: consumed,
      passes: config.passes,
      warmup_passes: config.warmupPasses,
      batch_size: config.arm === "materialized" ? null : config.batchSize,
      early_cancel: config.arm === "stream-cancel" ? config.earlyCancel : null,
      elapsed_ns: elapsedNs,
      checksum: Number(checksum),
    };
  } finally {
    dictionary.close();
  }
}

export const keyUnits = KEY_UNITS;
